### PR TITLE
Lps 172979

### PR DESCRIPTION
**Pull request by TaiPhamLR**
Motivation
=======
This work solves the problem reported in the ticket https://issues.liferay.com/browse/LPS-172979 .
I also added a unit test to cover the change. The test exposes the issue I mention below and makes sure the fix can cover the issue.
I also ran baseline.
Previous PR: https://github.com/liferay-core-infra/liferay-portal/pull/3546
(Please let me know if you have another solution).

Proposed Solution
========
The error occurs when we input `not ()` operand for filter. Ex: `not (tuition ge 123.46)`
* Cause:
   I found the root cause: The static method `not` at interface `Predicate` calls `predicate.not`. That makes a properly `rightExpression` and incorrect `leftExpression` on  new defaultPredicate.
After all, we will retrieve Predicate sql like this:  
`O_20096_Student_x.tuition_ >= ? not (O_20096_Student_x.tuition_ >=  ?)`
It's weird sql syntax. 

* What we did to solve this problem:
I create a `static class DoNothingExpression<T>` and method `notOf (Expression<Boolean> expression)` that return a new `DefaultPredicate` with the `leftExpression` is an instance of `DoNothingExpression`.  
The `DoNothingExpression` has an empty method `toSQL`.

Steps to Verify
========

Enable the feature flag: feature.flag.LPS-154672=true

1. Create Student object
2. Publish object
3. Create Student1, Student2
4. In Student create a field of type precision decimal 
5. Update Student1 with precision decimal value 123.45
6. Request: "${portalURL}/o/c/students?filter=(not (tuition ge 123.45)) or (tuition lt 0)"

### CODE REVIEW

File: modules/core/petra/petra-sql-dsl-spi/src/test/java/com/liferay/petra/sql/dsl/spi/SQLDSLTest.java [hunk 1]
```diff
@@ -977,6 +977,52 @@ public void testPredicateNot() {
...
+
+		Predicate notLtePredicate = Predicate.not(ltePredicate);
+
+		String informalQuery =
+			"MainExample.mainExampleId <= ? not (MainExample.mainExampleId " +
+				"<= ?) and MainExample.name = ? or MainExample.name = ?";
+
+		Assert.assertEquals(
+			informalQuery,
+			String.valueOf(notLtePredicate.and(defaultPredicate)));
```
Comment (TaiPhamLR) [inline modules/core/petra/petra-sql-dsl-spi/src/test/java/com/liferay/petra/sql/dsl/spi/SQLDSLTest.java:1001]: This assertion exposes that `Predicate.not` does not work correctly. It makes an informal query.

File: modules/core/petra/petra-sql-dsl-spi/src/test/java/com/liferay/petra/sql/dsl/spi/SQLDSLTest.java [hunk 2]
```diff
@@ -977,6 +977,52 @@ public void testPredicateNot() {
...
+			formalQueryNotOfAnd,
+			String.valueOf(notOfLtePredicate.and(defaultPredicate)));
+
+		String formalQueryAndNotOf =
+			"MainExample.name = ? or MainExample.name = ? and  not " +
+				"(MainExample.mainExampleId <= ?)";
+
+		Assert.assertEquals(
+			formalQueryAndNotOf,
+			String.valueOf(defaultPredicate.and(notOfLtePredicate)));
```
Comment (TaiPhamLR) [inline modules/core/petra/petra-sql-dsl-spi/src/test/java/com/liferay/petra/sql/dsl/spi/SQLDSLTest.java:1023]: These assertions expose that `Predicate.notOf` works well.

Comment (TaiPhamLR) [inline modules/core/petra/petra-sql-dsl-spi/src/test/java/com/liferay/petra/sql/dsl/spi/SQLDSLTest.java:1023]: These assertions expose that `Predicate.notOf` works well.

**Comment by catarineaguiar** [comment]
no regressions found.

**Comment by catarineaguiar** [comment]
ci:test:object

**Comment by zhangyan0701** [comment]
No regressions found.

**Comment by tinatian** [comment]
@TaiPhamLR 

From my last comment https://github.com/liferay-core-infra/liferay-portal/pull/3546#issuecomment-1435103510, I said I need an unit/integration test to expose the issue, not manual test step. I did not see the test I am expecting in this pull.

This should be a general issue which should be able to exposed by unit/integration test, the issue itself has nothing to do with this flag LPS-154672.

Please resend.

Thanks.
Tina.

**Comment by tinatian** [comment]
@TaiPhamLR Let me say it more clear. 

I need you to send me a pull, the first commit is a test, unit test or integration test, when I cherry pick your test to my local branch, and run it locally, I want to see it fail.

Flowing that commit, you modified the code to fix the issue.

Then I cherry pick you other commits and this time I want to see the failed test passed.

Please resend.

Thanks.
Tina.

**Comment by TaiPhamLR** [comment]
Thanks for your comments, and sorry for my misunderstanding.
I created a new PR: https://github.com/tinatian/liferay-portal/pull/2558
Tai.

**Pull request closed without merging**